### PR TITLE
main: refactor lmp-device-register

### PR DESCRIPTION
**Pull request by ldts**
Uses libp11 and the openssl API instead of the helper programs to access pkcs#11 and openssl.

Needs one fix to libp11 (for token initialization) and one missing feature (EC key generation, proposed upstream) plus a fix required to support TPM.

Co-developed-by: Milosz Wasilewski <milosz.wasilewski@foundries.io>

### CODE REVIEW

**Comment by ldts** [comment]
This is the first iteration of the refactoring.
From here we can improve encapsulation - or just write a proper C++ application if we want to.

**Comment by mike-sul** [comment]
I am getting the following compilation error when trying to build  on a local host

```
cmake --build build --target all
Consolidate compiler generated dependencies of target lmp-device-register
[ 16%] Building CXX object CMakeFiles/lmp-device-register.dir/src/main.cpp.o
In file included from /usr/include/openssl/cryptoerr_legacy.h:21,
                 from /usr/include/openssl/encodererr.h:17,
                 from /usr/include/openssl/encoder.h:21,
                 from /work/fio/prjs/lmp-device-register/inc/device_register.h:24,
                 from /work/fio/prjs/lmp-device-register/src/main.cpp:7:
/usr/include/openssl/macros.h:155:4: error: #error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
  155 | #  error "OPENSSL_API_COMPAT expresses an impossible API compatibility level"
      |    ^~~~~
make[2]: *** [CMakeFiles/lmp-device-register.dir/build.make:76: CMakeFiles/lmp-device-register.dir/src/main.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:83: CMakeFiles/lmp-device-register.dir/all] Error 2
make: *** [Makefile:136: all] Error 2

```

I guess that's because openssl `1.1.1s` is installed on my host while the given version supports only `3.x.x`?

**Comment by quaresmajose** [comment]
Maybe is better to require `OpenSSL 3.0.0` in the CMakeLists.txt
```
-find_package(OpenSSL REQUIRED)
+find_package(OpenSSL 3.0.0 REQUIRED)
```
https://github.com/foundriesio/lmp-device-register/blob/58efea0a9d107cf21ad6e72141698bb44227ee7b/CMakeLists.txt#L58

File: src/pkcs11.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+	std::regex UUID("("
+			"[a-f0-9]{8}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{4}-"
+			"[a-f0-9]{12}"
+			")");
+	std::smatch match;
+
+	for (int i = 0; i < nslots && opt.uuid.empty(); i++) {
```
Comment (mike-sul) [inline src/pkcs11.cpp:60]: My compiler prints warning for this.

```
/work/fio/prjs/lmp-device-register/src/pkcs11.cpp:60:27: warning: comparison of integer expressions of different signedness: 'int' and 'unsigned int' [-Wsign-compare]
   60 |         for (int i = 0; i < nslots && opt.uuid.empty(); i++) {

```

File: src/pkcs11.cpp [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+{
+	PKCS11_SLOT *slots = NULL;
+	PKCS11_SLOT *slot = NULL;
+	PKCS11_CTX *ctx = NULL;
+	PKCS11_KEY *key = NULL;
+	EVP_PKEY *prv = NULL;
+	EVP_PKEY *pub = NULL;
+	unsigned int nslots = 0;
+	unsigned int n = 0;
+	PKCS11_EC_KGEN ec = {
```
Comment (mike-sul) [inline src/pkcs11.cpp:143]: I get the following compilation error. Does this code depend on a specific libp11 version?

```
/work/fio/prjs/lmp-device-register/src/pkcs11.cpp: In function 'int pkcs11_create_csr(const lmp_options&, std::string&, std::string&)':
/work/fio/prjs/lmp-device-register/src/pkcs11.cpp:143:9: error: 'PKCS11_EC_KGEN' was not declared in this scope; did you mean 'PKCS11_TOKEN'?
  143 |         PKCS11_EC_KGEN ec = {
      |         ^~~~~~~~~~~~~~
      |         PKCS11_TOKEN

```

Comment (ldts) [reply src/pkcs11.cpp:143]: yes, please see this: https://github.com/foundriesio/meta-lmp/pull/1104/commits/804c6767fe50a60755e4e20e976d61f0ea6b3b96

Comment (quaresmajose) [reply src/pkcs11.cpp:143]: I see the same compilations errors:
```
error: ‘PKCS11_EC_KGEN’ was not declared in this scope; did you mean ‘PKCS11_TOKEN’?
error: ‘PKCS11_KGEN_ATTRS’ was not declared in this scope; did you mean ‘PKCS11_KEYGEN_FAILED’?
error: ‘attr’ was not declared in this scope
error: ‘ec’ was not declared in this scope
```

Comment (ldts) [reply src/pkcs11.cpp:143]: it integrates this better documented patch https://github.com/foundriesio/lmp-device-register/commit/ed399779f0576a0ad08c71e5aec8f055a202410b

File: CMakeLists.txt [hunk 3]
```diff
@@ -47,18 +47,21 @@ add_definitions('-g')
...
 # Use C++11, but without GNU or other extensions
 set(CMAKE_CXX_STANDARD 11)
 
-add_executable(lmp-device-register src/main.cpp)
-target_include_directories(lmp-device-register PRIVATE ${GLIB_INCLUDE_DIRS})
-target_link_libraries(lmp-device-register ${CURL_LIBRARIES} ${Boost_LIBRARIES} ${GLIB_LDFLAGS})
+add_executable(lmp-device-register src/main.cpp src/options.cpp src/auth.cpp src/pkcs11.cpp src/openssl.cpp)
+target_include_directories(lmp-device-register PRIVATE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/inc> ${GLIB_INCLUDE_DIRS})
+target_link_libraries(lmp-device-register ${CURL_LIBRARIES} ${Boost_LIBRARIES} ${GLIB_LDFLAGS} ${OPENSSL_LIBRARIES} ${LIBP11_LIBRARIES})
 install(TARGETS lmp-device-register RUNTIME DESTINATION bin)
```
Comment (quaresmajose) [inline CMakeLists.txt:67]: We can use the libp11.pc pkgconfig provide in the project. With this we can remove the `cmake-modules/FindLibP11.cmake`
```
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,17 +46,15 @@ message(STATUS "Setting SOTA_CLIENT to ${SOTA_CLIENT}")
 add_definitions('-g')
 add_definitions('-Wall')
 
-set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/cmake-modules)
-include_directories(SYSTEM ${LIBP11_INCLUDE_DIR})
-
 # find all required libraries
 set(Boost_USE_STATIC_LIBS ON)
 find_package(Boost COMPONENTS filesystem iostreams program_options system REQUIRED)
 find_package(CURL REQUIRED)
+find_package(OpenSSL 3.0.0 REQUIRED)
+
 find_package(PkgConfig REQUIRED)
-find_package(LibP11 REQUIRED)
-find_package(OpenSSL REQUIRED)
 pkg_search_module(GLIB REQUIRED glib-2.0)
+pkg_check_modules(LIBP11 REQUIRED libp11)
 
 # Use C++11, but without GNU or other extensions
 set(CMAKE_CXX_STANDARD 11)
```

File: src/pkcs11.cpp (hunk 2 above)
Comment (mike-sul) [reply src/pkcs11.cpp:143]: I can build the master branch that integrates that patch lmp-dev-reg on a local host (no HSM support) while I cannot build the given PR.

Comment (mike-sul) [reply src/pkcs11.cpp:143]: It looks like this commit https://github.com/foundriesio/meta-lmp/commit/804c6767fe50a60755e4e20e976d61f0ea6b3b96 is not merged into meta-lmp's main branch. How can I build this version of lmp-dev-reg then.

**Comment by mike-sul** [comment]
This version doesn't print help if no `/etc/os-release` is present.

```
./build/lmp-device-register --help
Can't read factory from /etc/os-release
Can't get the factory name

```

```
/build/lmp-device-register --factory msul-dev01 --help
Can't read factory from /etc/os-release
Can't get the factory name

```

**Comment by mike-sul** [comment]
```
DEVICE_FACTORY=msul-dev01 ./build/lmp-device-register --tags master
Read factory from environment: msul-dev01
Can't read tag from /etc/os-release
Can't get the tag information

```

**Comment by ldts** [comment]
> This version doesn't print help if no `/etc/os-release` is present.
> 
> ```
> ./build/lmp-device-register --help
> Can't read factory from /etc/os-release
> Can't get the factory name
> ```
> 
> ```
> /build/lmp-device-register --factory msul-dev01 --help
> Can't read factory from /etc/os-release
> Can't get the factory name
> ```

```
fio@intel-corei7-64:~$ ./lmp-device-register --help
lmp-device-register options:
  --help                                print usage
  --use-ostree-server arg (=1)          Use OSTree Proxy server instead of the 
                                        Device Gateway to pull the ostree repo
  -p [ --production ] arg (=0)          Enable a production release.
  --start-daemon arg (=1)               Start the aktualizr-lite systemd 
                                        service after registration.
  -d [ --sota-dir ] arg (=/var/sota)    The directory to install to keys and 
                                        configuration to.
  -g [ --device-group ] arg             Assign this device to a device group.
  -f [ --factory ] arg (=golem-x86)     The factory name to subscribe to.
  -S [ --hsm-so-pin ] arg               The PKCS#11 security officer pin - HSM 
                                        only
  -i [ --hwid ] arg (=intel-corei7-64)  The hardware identifier for the device 
                                        type
  -t [ --tags ] arg (=master)           Configure aktualizr-lite to only apply 
                                        updates from Targets with these tags.
  -T [ --api-token ] arg                API token for authentication. If not 
                                        provided, oauth2 will be used instead.
  -m [ --hsm-module ] arg               The PKCS#11 implementation (.so 
                                        library) - HSM only
  -P [ --hsm-pin ] arg                  The PKCS#11 pin - HSM only.
  -u [ --uuid ] arg                     A per-device UUID. If not provided, one
                                        will be generated. This is associated 
                                        with the device, e.g. as the CommonName
                                        field in certificates related to it.
  -n [ --name ] arg                     The name of the device as it should 
                                        appear in the dashboard. When not 
                                        specified, the device's UUID will be 
                                        used instead
  -H [ --api-token-header ] arg (=OSF-TOKEN)
                                        The HTTP header to use for 
                                        authentication.
  -a [ --apps ] arg                     Configure package-manager for this 
                                        comma separate list of apps.
  -A [ --restorable-apps ] arg (= )     Configure package-manager for this 
                                        comma separate list of Restorable Apps.
                                        If it is not specified, but a system 
                                        image is preloaded with Restorable Apps
                                        then the Restorable App list is set to 
                                        an empty list which means turning 
                                        restorable App usage ON and the 
                                        resultant list will be equal to the 
                                        apps list. Restorable App list = 
                                        UNION(compose-apps, restorable-apps)
Git Commit 4657c70007eae20418dfbe37b982d75b6840766d
```

**Comment by mike-sul** [comment]
>If /etc/os-release is not present it would print

lmp-dev-register should work even if `/etc/os-release` is missing.

**Comment by ldts** [comment]
> > If /etc/os-release is not present it would print
> 
> lmp-dev-register should work even if `/etc/os-release` is missing.

why?

**Comment by mike-sul** [comment]
> > > If /etc/os-release is not present it would print
> > 
> > 
> > lmp-dev-register should work even if `/etc/os-release` is missing.
> 
> why?

A few reasons:

1. Handle edge case: `/etc/os-release` is missing for unknown reason, removed, public LmP (no factory), etc.
2. `Factory` and `tags` can be specified via command line params what makes `/etc/os-release` redundant.
3. lmp-dev-reg is used for dev and performance testing in a few places; it is built on a regular linux host that does not have the release file, ubuntu, alpine, etc.

**Comment by ldts** [comment]
I pushed a change where lmp can take --os-release <file> from command line...failing that it will try /etc/os-release.

**Comment by doanac** [comment]
> BTW, I would use this refactoring as a chance to rename tags to tag since we don't allow devices to subscribe for many tags any more.

Let's be careful here. This commit isn't a refactor as we've already seen it changed how we dealt with a missing `os-release`. I'd say refactor or change functionality - but don't do both.

And if we change the CLI - we have to update docs and probably ought to keep a hidden option in place for a while to not break people doing this from muscle memory (or scripts).

**Comment by ldts** [comment]
> > BTW, I would use this refactoring as a chance to rename tags to tag since we don't allow devices to subscribe for many tags any more.
> 
> Let's be careful here. This commit isn't a refactor as we've already seen it changed how we dealt with a missing `os-release`. I'd say refactor or change functionality - but don't do both.
> 
> And if we change the CLI - we have to update docs and probably ought to keep a hidden option in place for a while to not break people doing this from muscle memory (or scripts).

right there is more than refactoring...the replacement of pkcs11-tool and openssl binaries.

**Comment by ldts** [comment]
the way IMO this should work:
first) read from configuration file (os-release or whatever input). But something must be set there.
then) read from command line - which will overwrite whatever was in the configuration file.

**Comment by ldts** [comment]
the questions is should a production system work without os_release.
if the answer is yes, sure, I will not make the configuration file mandatory
if the answer is no, then I will rather keep it mandatory for consistency.

**Comment by doanac** [comment]
> then I would remove --factory and --tags options and keep os-release for consistency with the production system.

tldr; We've been in circles on this and the past and wound up here.

But more broadly - this is unusual to see such a large change as a single commit that refactors the code base and changes functionality. I think you should just improve the code. Then we verify its a drop-in replacement. Then you can battle breaking compatibility in another commit where that can be disputed.

**Comment by ldts** [comment]
> > then I would remove --factory and --tags options and keep os-release for consistency with the production system.
> 
> tldr; We've been in circles on this and the past and wound up here.
> 
> But more broadly - this is unusual to see such a large change as a single commit that refactors the code base and changes functionality. I think you should just improve the code. Then we verify its a drop-in replacement. Then you can battle breaking compatibility in another commit where that can be disputed.

...when even parsing input options smells bad I think it is time to rewrite rather than to replicate the smell. Besides, the debug code should not impose on the production code....it should be seamless specially with the input params. this code has too much sedimentation IMO.

**Comment by mike-sul** [comment]
> @mike-sul try now. should be a proper refactor

It works  without the `os-release` file now.

File: src/pkcs11.cpp (hunk 2 above)
Comment (mike-sul) [reply src/pkcs11.cpp:143]: Maybe we should introduce some build flag so the patch is not required for non-lmp or non-HSM usage.
Ideally, it would be great to support softhsm without that patch if it's possible.

**Comment by mike-sul** [comment]
Is there any reasons not to print a device UUID instead of `Random`?

```
../build/lmp-device-register --factory msul-dev01 --tags master -d $PWD --start-daemon 0 -T `cat $FIO_TOKEN` -n "msul-test-device-00019965"
Can't read factory from /etc/os-release
Can't read tag from /etc/os-release
UUID: Random
Using DEVICE_API: https://api.foundries.io/ota/devices/
Registering device msul-test-device-00019965 with factory msul-dev01
Populate sota directory.
Device is now registered.

```

**Comment by ldts** [comment]
> Is there any reasons not to print a device UUID instead of `Random`?
> 
> 
> 
> ```
> 
> ../build/lmp-device-register --factory msul-dev01 --tags master -d $PWD --start-daemon 0 -T `cat $FIO_TOKEN` -n "msul-test-device-00019965"
> 
> Can't read factory from /etc/os-release
> 
> Can't read tag from /etc/os-release
> 
> UUID: Random
> 
> Using DEVICE_API: https://api.foundries.io/ota/devices/
> 
> Registering device msul-test-device-00019965 with factory msul-dev01
> 
> Populate sota directory.
> 
> Device is now registered.
> 
> 
> 
> ```

no  not really. shall we print it?

File: src/options.cpp [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+"The PKCS#11 pin - HSM only."
+
+static void get_factory_tags_info(const string os_release, string &factory,
+				  string &tag)
+{
+	const char *env = std::getenv(ENV_DEVICE_FACTORY);
+	ptree os_info;
+
+	if (env != nullptr) {
+		cout << "Factory read from environment" << endl
```
Comment (mike-sul) [inline src/options.cpp:87]: Semicolon is missing at the end of line.

File: src/pkcs11_stub.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+int pkcs11_store_cert(lmp_options &opt, X509 *cert)
+{
+	cout << "Executing PKCS11 stub " << endl;
+
+	return -1;
+}
+
+int int pkcs11_get_uuid(lmp_options &options)
```
Comment (mike-sul) [inline src/pkcs11_stub.cpp:23]: Just typo, `int` is typed twice.

File: src/pkcs11_stub.cpp [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+{
+	if (opt.hsm_module.empty())
```
Comment (mike-sul) [inline src/pkcs11_stub.cpp:25]:
```suggestion
	if (options.hsm_module.empty())
```

**Comment by mike-sul** [comment]
The current implementation doesn't support a custom registration server specified via `DEVICE_API` env variable. This is needed to support so-called "fully detached" and "partially detached" registration processes https://docs.foundries.io/latest/reference-manual/security/factory-registration-ref.html?highlight=device_api#fully-detached.

```
DEVICE_API=http://127.0.0.1/sign ./build/lmp-device-register --factory msul-dev01 --tag master -d $PWD --start-daemon 0 -T `cat $FIO_TOKEN`
Can't read tag from /etc/os-release
Can't read factory from /etc/os-release
UUID: a76ff6a4-9630-4292-8ba0-3f40081bf2aa [Random]
Setting factory name to UUID 
Using DEVICE_API: https://api.foundries.io/ota/devices/
Registering device a76ff6a4-9630-4292-8ba0-3f40081bf2aa with factory msul-dev01
Populate sota directory.
Device is now registered.
```

**Comment by ldts** [comment]
> The current implementation doesn't support a custom registration server specified via `DEVICE_API` env variable. This is needed to support so-called "fully detached" and "partially detached" registration processes https://docs.foundries.io/latest/reference-manual/security/factory-registration-ref.html?highlight=device_api#fully-detached.
> 
> ```
> DEVICE_API=http://127.0.0.1/sign ./build/lmp-device-register --factory msul-dev01 --tag master -d $PWD --start-daemon 0 -T `cat $FIO_TOKEN`
> Can't read tag from /etc/os-release
> Can't read factory from /etc/os-release
> UUID: a76ff6a4-9630-4292-8ba0-3f40081bf2aa [Random]
> Setting factory name to UUID 
> Using DEVICE_API: https://api.foundries.io/ota/devices/
> Registering device a76ff6a4-9630-4292-8ba0-3f40081bf2aa with factory msul-dev01
> Populate sota directory.
> Device is now registered.
> ```

just a bug. pushed the fix

File: src/pkcs11.cpp (hunk 2 above)
Comment (ldts) [reply src/pkcs11.cpp:144]: > Maybe we should introduce some build flag so the patch is not required for non-lmp or non-HSM usage. Ideally, it would be great to support softhsm without that patch if it's possible.

`-DDISABLE_PKCS11=ON ` to disable building-in pkcs11 support

File: CMakeLists.txt (hunk 3 above)
Comment (ldts) [reply CMakeLists.txt:73]: that would be great. let me test this.

File: src/pkcs11.cpp (hunk 2 above)
Comment (mike-sul) [reply src/pkcs11.cpp:144]: It works, thanks.

**Comment by ldts** [comment]
added `--mlock-all <bool>` as optional for development.

**Review by mike-sul — COMMENTED** [review]
LGTM

**Comment by ldts** [comment]
@ricardosalveti @mike-sul I added the missing support for 'production' rollouts

File: src/options.cpp [hunk 7]
```diff
@@ -65,7 +65,7 @@ namespace po = boost::program_options;
 "Assign this device to a device group."
 
 #define PRODUCTION_HELP \
-"Enable a production release."
+"Allow pushed updates (production updates) on the device."
```
Comment (vkhoroz) [inline src/options.cpp:68]: I think this new help message takes it somewhat sidelines.

The PRODUCTION flag does not enable pushed updates.
What it does is: designate any devices as production devices when they are registered using a version of `lmp-dev-reg` built with this flag. That way these devices get a completely different subset of updates from regular CI (test) devices. These are not "pushed updates" - they are simply "production updates".

IIUC this option is then enabled on a per-branch level in the factory using this switch:
https://github.com/foundriesio/meta-lmp/blob/main/meta-lmp-base/recipes-sota/lmp-device-register/lmp-device-register_git.bb#L15

I think @doanac or @mike-sul can help more here to understant how this plays out with the factory's CI configs.

Comment (vkhoroz) [reply src/options.cpp:68]: Nevertheless, I'm not sure I'd like to mention "production updates" explicitly here.

Right now the only difference between the test vs production device is indeed the set of updates they get.
But, that difference can be extended in the future.
Also, our users are free to extend that difference right now, as they can read and react to this attribute.

Therefore, I'd prefer to keep it more flexible like this: `Designate|Mark the device as a production device.`

Comment (doanac) [reply src/options.cpp:68]: I don't think anyone fixes it on a per-branch setting. They leave it as a run-time option for devices. I guess the help message should be something like: "flag device to receive updates via waves"

Comment (ldts) [reply src/options.cpp:68]: I understood the main difference to be that the device didnt need to ping periodically for its updates but that the update to take would be pushed instead. is this not the case? how shall I reword the option then?

Comment (vkhoroz) [reply src/options.cpp:68]: I think one of these options are OK:
- `Mark the device as a production device.` - this is my preference.
- `Enable production updates on the device.` - this gives a more exact context, but already too many details IMO.

Comment (vkhoroz) [reply src/options.cpp:68]: IIRC we operate a term "production device" extensively in our docs/blogs by now.
So, it should be clear enough to a user what it means.

**Comment by ldts** [comment]
goint to fold all the commits into a single one

**Review by ricardosalveti — APPROVED** [review]
Lets merge this one and get it in LmP for wider testing.
